README.md
![](gallery.png)

`vegalite` : Do whatever `ggvis` doesn't :-)

For better examples, [take a look here](http://rud.is/projects/vegalite01.html).

There's also a [blog post](http://rud.is/b/2016/02/27/create-vega-lite-specs-widgets-with-the-vegalite-package/) describing this in a bit more detail.

The vegalite package has been updated to Vega-Lite 2.0, but is not yet 100% feature-complete.

In addition to functions for making a Vega-Lite based htmlwidget, functions also exist which enable creation of widgets from a JSON spec and turning a `vegalite` package created object into a JSON spec.

You start by calling `vegalite()` which allows you to setup core configuration options, including whether you want to display links to show the source and export the visualization. You can also set the background here and the `viewport_width` and `viewport_height`. Those are very important as they control the height and width of the widget and also the overall area for the chart. This does _not_ set the height/width of the actual chart. That is done with `view_size()`.

Once you instantiate the widget, you need to `add_data()` which can be `data.frame`, local CSV, TSV or JSON file (that convert to `data.frame`s) or a non-realive URL (wich will not be read and converted but will remain a URL in the Vega-Lite spec.

You then need to `encode_x()` & `encode_y()` variables that map to columns in the data spec and choose one `mark_...()` to represent the encoding.

Here's a sample, basic Vega-Lite widget:

    dat <- jsonlite::fromJSON('[
        {"a": "A","b": 28}, {"a": "B","b": 55}, {"a": "C","b": 43},
        {"a": "D","b": 91}, {"a": "E","b": 81}, {"a": "F","b": 53},
        {"a": "G","b": 19}, {"a": "H","b": 87}, {"a": "I","b": 52}
      ]')

    vegalite() %>%
      add_data(dat) %>%
      encode_x("a", "ordinal") %>%
      encode_y("b", "quantitative") %>%
      mark_bar() -> vl

     vl

That is the minimum set of requirements for a basic Vega-Lite spec and will create a basic widget.

You can also convert that R widget object `to_spec()` which will return the JSON for the Vega-Lite spec (allowing you to use it outside of R).

    to_spec(vl)

    {
      "description": "",
      "data": {
        "values": [
          { "a": "A", "b": 28 }, { "a": "B", "b": 55 }, { "a": "C", "b": 43 },
          { "a": "D", "b": 91 }, { "a": "E", "b": 81 }, { "a": "F", "b": 53 },
          { "a": "G", "b": 19 }, { "a": "H", "b": 87 }, { "a": "I", "b": 52 }
        ]
      },
      "mark": "bar",
      "encoding": {
        "x": {
          "field": "a",
          "type": "nominal"
        },
        "y": {
          "field": "b",
          "type": "quantitative"
        }
      },
      "config": [],
      "embed": {
        "renderer": "svg",
        "actions": {
          "export": false,
          "source": false,
          "editor": false
        }
      }
    }

If you already have a Vega-Lite JSON spec that has embedded data or a non-realtive URL, you can create a widget from it via `from_spec()` by passing in the full JSON spec or a URL to a full JSON spec.

If you're good with HTML (etc) and want a more lightweight embedding options, you can also use `embed_spec` which will scaffold a minimum `div` + `script` source and embed a spec from a `vegalite` object.

If you like the way Vega-Lite renders charts, you can also use them as static images in PDF knitted documents with the new `capture_widget` function. (NOTE that as of this writing, you can just use the development version of `knitr` instead of this function.)

### Installation


```r
devtools::install_github("hrbrmstr/vegalite")
```



### Usage


```r
library(vegalite)

# current verison
packageVersion("vegalite")
```

```
## [1] '2.0.0.9999'
```


```r
library(vegalite)

vegalite() %>% 
  cell_size(400, 400) %>% 
  add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>% 
  encode_x("Horsepower") %>% 
  encode_y("Miles_per_Gallon") %>% 
  encode_color("Origin", "nominal") %>% 
  mark_point()
```

![](vega.png)

### Code of Conduct

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.



NEWS
## 2.0.1
* add support for rendering a spec with shiny

## 2.0.0
* Initial support for vega-lite 2.0
* Updates Javascript libraries for upgrade to vega-lite 2.0. New versions of vega-lite, vega, vega-embed, vega-tooltip. Remove separate D3 library because not necessary.
* Updates spec.js and vegalite.js to match new api for vega-embed and vega-tooltip.
* Add z-schema javascript library and a function validate_vl that checks whether schema is valid vega-lite2. Updates tests to use validate_vl.
* Updates to to_spec and embed_spec to give valid spec for vega-lite2
* Deprecate axis_facet_row and axis_facet_col because behaviour currently un-supported by vega-lite 2.
* Update to axis_vl function so that arguments match those supported by vega-lite 2 axis config.
* band_size argument now named range_step, to match vega-lite 2 naming
* cell_size and cell_config deprecated in favor of view_size and view_config to better match new naming of field in vega-lite 2
* grid_facet deprecated because there is no longer a facet field in config in vega-lite 2
* Bug fixes to capture_widget


## 0.7.1
* Adds _vl to places where vegalite is likely to collide with other common
name spaces.

## 0.7.0
* Updates vega-lite to final version before vega-lite 2
* Refactors each of the functions to have a general form like mark() that the specific forms, like mark_point() call. This means that new options can be added in one place and not in each iteration of the same function.
* All new arguments have been added.
* Documentation for all the new arguments.
* Light test coverage based on the vega*.json examples.

## 0.6.1.9000
* changed the way `%>%` is exported

## 0.6.0.9000
* handles ts objects
* ensures passed-in data.frame-like object is converted to a data.frame

## 0.5.1.9000
* Widget to png for PDF use

## 0.5.0.9000
* More examples. Passes R-devel CRAN pre-flight check

## 0.4.0.9000
* more config_ functions and better parameters for mark_'s

## 0.3.5.9000
* passes CRAN checks (which cleans up alot of things)

## 0.3.0.9000
* Legends, sorting, scales, order, detail

## 0.2.0
* Documentation

## 0.1.0
* Initial release


R/mark.r
#' Mark
#'
#' A bar mark represents each data point as a rectangle, where the length is
#' mapped to a quantitative scale.
#'
#' @param vl Vega-Lite object
#' @param mark can be "bar", "circle", "square", "tick", "line", "area", "point",
#'        and "text". These directly set how the data are drawn and are similar
#'        geoms in ggplot2.
#' @param filled,color,fill,stroke see \href{https://vega.github.io/vega-lite/docs/config.html#color}{config.mark color docs}
#' @param opacity,fillOpacity,strokeOpacity see \href{https://vega.github.io/vega-lite/docs/config.html#opacity}{config.mark opacity docs}
#' @param strokeWidth,strokeDash,strokeDashOffset see \href{https://vega.github.io/vega-lite/docs/config.html#stroke-style}{config.mark stroke docs}
#' @param stacked Defunct; Use in \code{\link{encode_x}} or \code{\link{encode_y}}
#' @param interpolate,tension for line and area \code{mark}, the line interpolation method.
#'        value for interpolate can be "linear", "step-before", "step-after",
#'        "basis", "basis-open", "basis-closed", "bundle", "cardinal",
#'        "cardinal-open", "cardinal-closed", or "monotone", For tension, see
#'        \href{https://github.com/mbostock/d3/wiki/SVG-Shapes#line_interpolate}{D3's line interpolation}.
#' @param orient the orientation of a non-stacked bar, area, and line charts.
#'        The value is either "horizontal", or "vertical" (default). For bar and
#'        tick, this determines whether the size of the bar and tick should be
#'        applied to x or y dimension. For area, this property determines the
#'        orient property of the Vega output. For line, this property determines
#'        the path order of the points in the line if path channel is not specified.
#'        For stacked charts, this is always determined by the orientation of the stack;
#'        therefore explicitly specified value will be ignored.
#' @param barSize the size of the bars (width or height depending on \code{orient})
#' @param shape applicable to point \code{mark}, "circle", "square", "cross", "diamond",
#'        "triangle-up", or "triangle-down", or else a custom SVG path string.
#' @param size for point, circle or square \code{mark}, the pixel area of a point.
#' @param tickSize the size of ticks
#' @param tickThickness the thickness of ticks.
#' @param ... additional arguments passed to mark
#' @rdname mark
#' @encoding UTF-8
#' @references \href{http://vega.github.io/vega-lite/docs/mark.html}{Vega-Lite Mark spec},
#' \href{https://vega.github.io/vega-lite/docs/config.html#mark-config}{Vega-Lite config.mark spec}
#' @export
#' @examples
#' dat <- jsonlite::fromJSON('[
#'     {"a": "A","b": 28}, {"a": "B","b": 55}, {"a": "C","b": 43},
#'     {"a": "D","b": 91}, {"a": "E","b": 81}, {"a": "F","b": 53},
#'     {"a": "G","b": 19}, {"a": "H","b": 87}, {"a": "I","b": 52}
#'   ]')
#'
#' vegalite() %>%
#'   add_data(dat) %>%
#'   encode_x("a", "ordinal") %>%
#'   encode_y("b", "quantitative") %>%
#'   mark_bar()
mark <- function(vl, mark="circle", filled=NULL, color=NULL, fill=NULL,
                 stroke=NULL, opacity=NULL, fillOpacity=NULL,
                 strokeOpacity=NULL, strokeWidth=NULL, strokeDash=NULL,
                 strokeDashOffset=NULL, stacked=NULL, interpolate=NULL,
                 tension=NULL, orient=NULL, barSize=NULL, shape=NULL, size=NULL,
                 tickSize=NULL, tickThickness=NULL) {

  vl$x$mark <- mark

  if (!is.null(filled))           vl$x$config$mark$filled <- filled
  if (!is.null(color))            vl$x$config$mark$color <- color
  if (!is.null(fill))             vl$x$config$mark$fill <- fill
  if (!is.null(stroke))           vl$x$config$mark$stroke <- stroke
  if (!is.null(opacity))          vl$x$config$mark$opacity <- opacity
  if (!is.null(fillOpacity))      vl$x$config$mark$fillOpacity <- fillOpacity
  if (!is.null(strokeOpacity))    vl$x$config$mark$strokeOpacity <- strokeOpacity
  if (!is.null(strokeWidth))      vl$x$config$mark$strokeWidth <- strokeWidth
  if (!is.null(strokeDash))       vl$x$config$mark$strokeDash <- strokeDash
  if (!is.null(strokeDashOffset)) vl$x$config$mark$strokeDashOffset <- strokeDashOffset

  if (mark %in% c("bar", "area")) {
    if (!is.null(stacked)) {
      warning("stacked argument in mark will be ignored.",
              "Use in encode_* instead")
    }
  }

  if (mark %in% c("line", "area")){
    if (!is.null(interpolate))    vl$x$config$mark$interpolate <- interpolate
    if (!is.null(tension))        vl$x$config$mark$tension <- tension
  }

  if (mark %in% c("bar", "area", "line")) {
    if (!is.null(orient))         vl$x$config$mark$orient <- orient
  }

  if (mark == "bar") {
    if (!is.null(barSize))        vl$x$config$mark$barSize <- barSize
  }
  if (mark == "point") {
    if (!is.null(shape))          vl$x$config$mark$shape <- shape
  }

  if (mark %in% c("point", "circle", "square")) {
    if (!is.null(size))           vl$x$config$mark$size <- size
  }

  if (mark == "tick"){
    if (!is.null(tickSize))       vl$x$config$mark$tickSize <- tickSize
    if (!is.null(tickThickness))  vl$x$config$mark$tickThickness <- tickThickness
  }

  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Miles_per_Gallon", "quantitative") %>%
#'   mark_circle()
mark_bar <- function(vl, ...) {
  vl <- mark(vl, mark="bar", ...)
  return(vl)
}
#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Miles_per_Gallon", "quantitative") %>%
#'   mark_circle()
mark_circle <- function(vl, ...) {
  vl <- mark(vl, mark="circle", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Miles_per_Gallon", "quantitative") %>%
#'   mark_square()
mark_square <- function(vl, ...) {
  vl <- mark(vl, mark="square", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Cylinders", "ordinal") %>%
#'   mark_tick()
mark_tick <- function(vl, ...){
  vl <- mark(vl, mark="tick", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   view_size(300, 300) %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/driving.json") %>%
#'   encode_x("miles", "quantitative") %>%
#'   encode_y("gas", "quantitative") %>%
#'   encode_order("year", "temporal") %>%
#'   scale_x_linear_vl(zero=FALSE) %>%
#'   scale_y_linear_vl(zero=FALSE) %>%
#'   mark_line()
mark_line <- function(vl, ...) {
  vl <- mark(vl, mark="line", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   view_size(300, 200) %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/unemployment-across-industries.json") %>%
#'   encode_x("date", "temporal") %>%
#'   encode_y("count", "quantitative", aggregate="sum") %>%
#'   encode_color("series", "nominal") %>%
#'   scale_color_nominal_vl(scheme="category20b") %>%
#'   timeunit_x("yearmonth") %>%
#'   scale_x_time_vl(nice="month") %>%
#'   axis_x(format="%Y", labelAngle=0) %>%
#'   mark_area()
mark_area <- function(vl, ...) {
  vl <- mark(vl, mark="area", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Miles_per_Gallon", "quantitative") %>%
#'   mark_point()
mark_point <- function(vl, ...) {
  vl <- mark(vl, mark="point", ...)
  return(vl)
}

#' @rdname mark
#' @export
#' @examples
#' vegalite() %>%
#'   view_size(300, 200) %>%
#'   add_data("https://vega.github.io/vega-editor/app/data/cars.json") %>%
#'   encode_x("Horsepower", "quantitative") %>%
#'   encode_y("Miles_per_Gallon", "quantitative") %>%
#'   encode_color("Origin", "nominal") %>%
#'   calculate("OriginInitial", "datum.Origin[0]") %>%
#'   encode_text("OriginInitial", "nominal") %>%
#'   mark_text()
mark_text <- function(vl, ...) {
  vl <- mark(vl, mark="text", ...)
  return(vl)
}
